#11 Feat : 회원가입 - 아이디, 비밀번호, 이메일 유효성 검사

uri010 opened the pull request:
## 요약
- 아이디, 비밀번호, 이메일 유효성 검사 기능 추가

## 변경 내용
- 유효성 검사 클래스를 사용하기 위해 build.gradle에 의존성을 추가합니다. 
- 회원가입시 아이디와 이메일은 User 클래스에서 유효성을 검사합니다.
    - MethodArgumentNotValidException 예외를 ExceptionHandler로 처리합니다.
- 회원가입시 비밀번호는 passwordValidation 메소드를 통해 유효성을 검사합니다.
- 아이디 중복 확인시 아이디는 userIdValidation 메소드를 통해 유효성을 검사합니다.
- 회원가입과 아이디 유효성을 확인하는 메소드에서 발생하는 ValidationException 예외를 ExceptionHandler로 처리합니다. 

## Issue number/URL
#11 Feat : 회원가입 - 아이디, 비밀번호, 이메일 유효성 검사

uri010 committed:
#11 feat : 아이디, 비밀번호, 이메일 유효성 검사 기능 추가

 유효성 검사 클래스를 사용하기 위해 build.gradle에 의존성 추가
 회원가입시 아이디와 이메일은 User 클래스에서 유효성을 검증하고, 비밀번호는 따로 검증 후 암호화
 아이디 중복 확인시 아이디 유효성을 따로 검증하는 메소드 추가
 User 클래스의 유효성 검사의 MethodArgumentNotValidException예외 처리
 아이디 중복 확인시 유효성 검사의 ValidationException 예외 처리

sunwootest commented on src/main/java/study/sharedcalendar/service/UserService.java:
```diff
@@ -39,4 +44,25 @@ public void userIdDuplicationCheck(String userId) {
         return userMapper.userIdExist(userId);
     }
+
+    public void passwordValidation(String password) {
+        String passwordPolicy = "^(?=.*[a-z])(?=.*[A-Z])(?=.*\\d)(?=.*[$@$!%*?&])[A-Za-z\\d$@$!%*?&]{8,15}$";
+
+        Pattern pattern = Pattern.compile(passwordPolicy);
+        Matcher matcher = pattern.matcher(password);
+        if (!matcher.matches()) {
+            throw new ValidationException("잘못된 비밀번호 형식입니다.");
+        }
+    }
+
+    public void userIdValidation(String userId) {
+        String userIdPolicy = "^[0-9a-zA-Z]{3,10}$";
+
+        Pattern pattern = Pattern.compile(userIdPolicy);
+        Matcher matcher = pattern.matcher(userId);
+
+        if (!matcher.matches()) {
+            throw new ValidationException("잘못된 아이디 형식입니다.");
+        }
+    }
```
presentation layer 뿐만 아니라 business layer에도 검증 로직을 추가하셨는데 이유가 있을까요? 이렇게 구현할 경우 장단점도 알아보시면 좋을 것 같네요 :)

sunwootest replied:
springboot의 validator의 구조에 대해서 알아보시면 좋을 것 같습니다. MemberService에 검증이라는 관심사가 생겼네요 :)

uri010 replied:
비밀번호의 경우 Pattern 어노테이션으로 요청으로 보낸 값에는 유효성 검사를 할 수 있지만, 암호화를 한 비밀번호 값을 builder로 User 객체를 생성하게 되면 그때에도 Pattern 유효성 검사가 이루어져서 예외가 발생했습니다. 
결국 선택한 게 유효성을 따로 해주는 방법이었는데 좀 더 공부해보겠습니다.

uri010 committed:
유효성 검사를 UserController에서 처리
Validated 유효성 검사 실패시 ContraintViolationException 예외 처리

uri010 commented:
## 요약
- 유효성 검사 Presentation Layer에서 처리

## 변경 내용
- 유효성 검사를 UserController에서 처리해줍니다.
- Validated 유효성 검사 실패시 ConstraintViolationException 예외를 처리해줍니다. 

## 참고
- https://reflectoring.io/bean-validation-with-spring-boot/

sunwootest commented on src/main/java/study/sharedcalendar/controller/UserController.java:
```diff
@@ -1,23 +1,32 @@
-    public void idCheck(@RequestParam String userId) {
+    public void idCheck(
+            @RequestParam
+            @Pattern(message = "잘못된 아이디 형식입니다."
+                    , regexp = "^[a-z0-9_-]{3,10}")
+            String userId) {
```
이렇게 작성하신 이유가 있을까요?

uri010 replied:
참고에 올려놓은 페이지에서 유효한 값이 아닌데 비즈니스 로직으로 넘어가는것 보다 유효성을 체크한 뒤에 비즈니스 로직으로 넘어가는게 좋다는 의견이 타당하다고 생각해서 유효성 검사를 컨트롤러에서 처리했습니다.

sunwootest reviewed (commented):
비즈니스 로직에 대한 유효성 검증을 아예 삭제하신 것으로 확인됩니다. 이전 리뷰에 남긴 내용은 @Valid와 @Validated에 대해서 알아보라는 의도에서 작성했습니다 :)

uri010 committed:
#11 fix : 회원가입 DTO 클래스 User를 SignUpReq로 수정 및 속성 수정

DTO의 특성을 나타내기 위해 클래스 이름을 SignUpReq로 수정
SignUpReq의 속성 중 id와 inviteUrl은 요청을 받을 때 필요하지 않는 속성이므로 삭제

uri010 commented:
## 요약
- User를 SignUpReq로 수정

## 변경 내용
- DTO의 특성을 나타내기 위해 클래스 이름을 SignUpReq로 수정합니다.
- SignUpReq의 속성 중 id와 inviteUrl은 요청을 받을 때 필요하지 않는 속성이므로 삭제합니다.

sunwootest commented:
임시 approve 드립니다!
